### ISSUE TITLE
LWT: LWT write contention occurs on one shard, will it affect other shards?

### ISSUE THREAD

**Issue by Felix-zhoux** [scylladb/scylla#7547]
*Installation details*
Scylla version (or git commit hash): 4.1.7
Cluster size: 18
OS (RHEL/CentOS/Ubuntu/AWS AMI): CentOS

Hi, I have a performance problem:
In my data model, there is a frequent UpdateItem operation for a single item (this logic is incorrect and will be optimized later), which means that there is a shard (processing this specific partition) that will be very busy, busy processing LWT UpdateItem. Therefore, as shown in the figure below, this **shard 28** has LWT write contention. This is within my expectations.

![image](https://user-images.githubusercontent.com/59597542/98087900-bdfb1b00-1ebb-11eb-9cca-169d4631715d.png)

However, during the test, it was found that the shards handling PutItem operations of other items (different tables, different partitions) occasionally experienced excessive delays. From the perspective of LWT monitoring, the delay seems to be spent on processing LWT. As shown in the figure above, **shard 3** handles normal PutItem requests. Is this situation affected by the above UpdateItem operation?

In other words, the occurrence of LWT contention on one shard will affect the LWT operations of other shards?

**Comment by nyh** [comment]
I have a guess that might explain why you can see high tail latencies on requests submitted to uncontended items and other shards as well. CC @gleb-cloudius who's the expert in this and may support or dispute my guess:

Although each operation on a specific partition will **eventually** be handled by a specific shard, the request is originally received on some other random shard, say shard 3 in your example, and then needs to be "bounced" to the correct shard. This bouncing happens through a queue (`alternator::executor::_ssg`) which has a fixed capacity for each destination shard. Because destination shard 28 will receive a lot of **high-latency** operations (that end up waiting a long time because of contention), the queue going from shard 3 to shard 28 can become full with these hung operations - at which point all new requests on shard 3 that need to go to 28 - and not just to the contended item but also to other random items - will hang. Basically, if you have 32 shards, as much as 3% arriving at shard 3 will need to be forwarded to shard 28, and all those 3% will get delayed and result in a high 97th percentile (and obviously 99.9th percentile) delay.
 
If my guess is correct, you might be able to improve the situation by increasing the queue size, defined in `main.cc` in the code:
```c++
                // Create an smp_service_group to be used for limiting the
                // concurrency when forwarding Alternator request between
                // shards - if necessary for LWT.
                smp_service_group_config c;
                c.max_nonlocal_requests = 5000;
                smp_service_group ssg = create_smp_service_group(c).get0();
```
Try to increase this number "5000" significantly, and see if things improve.

**Comment by gleb-cloudius** [comment]
What is missing here is the shard load graph. Can it be that shard 28 is
overloaded?

--
			Gleb.

**Comment by Felix-zhoux** [comment]
> What is missing here is the shard load graph. Can it be that shard 28 is overloaded?
> […](#)
> -- Gleb.

As I see, yes! This is a typical situation. In fact, shard 28 may also process these PutItem requests.
![image](https://user-images.githubusercontent.com/59597542/98096712-0e2baa80-1ec7-11eb-8d2d-d572d1262318.png)
![image](https://user-images.githubusercontent.com/59597542/98097247-b6da0a00-1ec7-11eb-9b72-c110b2e6f696.png)

**Comment by Felix-zhoux** [comment]
> What is missing here is the shard load graph. Can it be that shard 28 is overloaded?
> […](#)
> -- Gleb.

As I see, yes! This is a typical situation. In fact, shard 28 may also process these PutItem requests.

![image](https://user-images.githubusercontent.com/59597542/98097642-3c5dba00-1ec8-11eb-8ddf-a24a3435c464.png)

![image](https://user-images.githubusercontent.com/59597542/98097661-47184f00-1ec8-11eb-8447-0d9b8c560f3b.png)

**Comment by slivne** [comment]
@Felix-zhoux - are you spreading the requests on all shards ? requests should spread out between the shards 

@gleb-cloudius ?  anything else to check

**Comment by gleb-cloudius** [comment]
I am not sure what else can be said here. All requests whose data resides on shard 28 will have bad latency due to shard been overloaded.

**Comment by Felix-zhoux** [comment]
I think @nyh nhy's guess has answered this question. If this guess is  correct, the shard utilization rate in the screenshot also verifies this guess.
